larreco/ClusterFinder/ClusterAna_module.cc

```diff
@@ -115,611 +138,471 @@ namespace cluster {
-                }
-              } // correct view
-            }   // ii
-            mfp << plane << ":" << loW << ":" << loT << "-" << plane << ":" << hiW << ":" << hiT
-                << " ";
-          } // plane
-        }   // fPrintLevel > 2
-      }     // !isCosmic
-    }       // ipl
-
-    float ave1 = -1.;
-    if (numNuEP2mu > 0.) ave1 = aveNuEP2mu / numNuEP2mu;
+        if (effpur < 0.7) ++fNBadEP;
+        if (fPrintLevel > 0 && effpur < 0.7) {
+          mf::LogVerbatim myprt("ClusterAna");
+          myprt << "BadEP evt " << evt.event();
+          myprt << " " << fNames[indx];
+          if (match.clsIndex[plane] != UINT_MAX) {
+            auto& cls = (*allCls)[match.clsIndex[plane]];
+            myprt << " clsID " << cls.ID();
+          }
+          myprt << " EP " <<std::fixed << std::setprecision(2) << effpur;
+          unsigned int first, last;
+          FindFirstLastWire(allHits, hitMCPIndex, match, plane, first, last);
+          if (first != UINT_MAX) {
+            auto& fhit = (*allHits)[first];
+            auto& lhit = (*allHits)[last];
+            myprt << " true hit range " << PrintHit(fhit) << " - " << PrintHit(lhit);
+            myprt << " nTrueHits " << (int)match.mcpTruHitCount[plane];
+          }
+        } // fPrintLevel > 0 && effpur < 0.7
+      } // plane
+    } // match
 
-    float ave2 = -1.;
-    if (numNuEP2nm > 0.) ave2 = aveNuEP2nm / numNuEP2nm;
+  } // analyze
 
-    float ave3 = -1.;
-    if (numCREP2 > 0.) ave3 = aveCREP2 / numCREP2;
+  void
+  ClusterAna::FindFirstLastWire(art::Handle<std::vector<recob::Hit>> const& allHits,
+                                std::vector<unsigned int> const& hitMCPIndex,
+                                MatchStruct const& match,
+                                unsigned short inPlane, unsigned int& first, unsigned int& last)
+  {
+    // Returns the index of the first hit and last hit in the specified plane
+    // where first (last) = the lowest (highest) wire number
+    first = UINT_MAX;
+    last = 0;
+    if (match.mcpi == UINT_MAX) return;
+    unsigned int fWire = UINT_MAX;
+    unsigned int lWire = 0;
+    for(unsigned int iht = match.firstHit[inPlane]; iht <= match.lastHit[inPlane]; ++iht) {
+      if (hitMCPIndex[iht] != match.mcpi) continue;
+      auto& hit = (*allHits)[iht];
+      if (hit.WireID().Plane != inPlane) continue;
+      if (hit.WireID().TPC != match.tpc) continue;
+      unsigned int wire = hit.WireID().Wire;
+      if (wire < fWire) {
+        fWire = wire;
+        first = iht;
+      } // wire > fWire
+      if (wire > lWire) {
+        lWire = wire;
+        last = iht;
+      } // wire > fWire
+    } // iht
+  } // FindFirstLastWire
+
+  int
+  ClusterAna::PDGCodeIndex(int pdgCode)
+  {
+    pdgCode = abs(pdgCode);
+    if (pdgCode == 11) return 0;
+    if (pdgCode == 13) return 1;
+    if (pdgCode == 211) return 2;
+    if (pdgCode == 321) return 3;
+    if (pdgCode == 2212) return 4;
+    return -1;
+  }
```

Comment:
As suggested above, this would be replaced by a function that returns an enumeration instead of an integer.